### PR TITLE
builds: update all GH action workflows to use latest versions

### PR DESCRIPTION
**Pull request by deadprogram**
This PR updates all GH action workflows to use latest versions of shared workflows.

### CODE REVIEW

**Comment by sago35** [comment]
Here are the links to the documentation. 

V4 https://github.com/actions/checkout
V3 https://github.com/actions/upload-artifact
V4 https://github.com/actions/setup-go
V3 https://github.com/docker/setup-buildx-action
V5 https://github.com/docker/metadata-action
V3 https://github.com/docker/login-action
V5 https://github.com/docker/build-push-action

**Comment by deadprogram** [comment]
@sago35 I thought this PR updated to those versions you mention. DId I miss something?

**Comment by sago35** [comment]
I'm just creating the documentation links for my understanding. @deadprogram

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -19,15 +19,23 @@ jobs:
...
       contents: read
     steps:
+      - name: Free Disk space
+        shell: bash
+        run: |
+          sudo rm -rf /usr/local/lib/android
+          sudo rm -rf /usr/share/dotnet
+          sudo rm -rf /opt/ghc
+          sudo rm -rf /usr/local/graalvm
+          sudo rm -rf /usr/local/share/boost
```
Comment (sago35) [inline .github/workflows/docker.yml]: I think it would be better to make this change in a separate PR if possible.

Comment (deadprogram) [reply .github/workflows/docker.yml]: That is what is in https://github.com/tinygo-org/tinygo/pull/4009

Comment (deadprogram) [reply .github/workflows/docker.yml]: @sago35 sorry, I misunderstood your comment. I will create a separate PR.

Comment (deadprogram) [reply .github/workflows/docker.yml]: See https://github.com/tinygo-org/tinygo/pull/4013

**Review by sago35 — APPROVED** [review]
LGTM

**Comment by deadprogram** [comment]
Thanks for review @sago35 now merging.
